File: src/main/java/com/ballgame/Fireball.java
package com.ballgame;

import java.awt.Graphics2D;
import java.awt.geom.Ellipse2D;
import java.awt.Shape;
import java.awt.image.*;
import java.util.List;

public class Fireball extends Circle {
    private Shape shape;
    private double velX;
    private double velY;

    public Fireball(List<BufferedImage> sprites) {
        this(0, 0, 0, 80, 80, sprites);
    }

    public Fireball(int x, int y, double velX, double velY, int r, List<BufferedImage> sprites) {
        super(x, y);

        this.initSprites(sprites, 15, 1.6);

        this.velX = velX;
        this.velY = velY;
        this.r = r;

        this.shape = new Ellipse2D.Double((int) this.x, (int) this.y, this.getDiameter(), this.getDiameter());
    }

    public Ellipse2D.Double getShape() {
        return (Ellipse2D.Double) this.shape;
    }

    public double getVelX() {
        return this.velX;
    }

    public double getVelY() {
        return this.velY;
    }

    public void setVelX(double velX) {
        this.velX = velX;
    }

    public void setVelY(double velY) {
        this.velY = velY;
    }

    public void rotateSprites() {
        double angle = Math.atan(this.velX / -this.velY);
        for (int i = 0; i < this.sprites.size(); i++) {
            BufferedImage rotatedImage = this.rotateImage(this.sprites.get(i), angle * 180);
            this.transformedSprites.set(i, rotatedImage);
        }
    }

    public void update(double deltaTime, double previousTime) {
        boolean incrementedFrame = this.incrementImageFrame(deltaTime, previousTime);

        if (incrementedFrame) {
            // double angle = Math.atan(this.velX / -this.velY);
            //BufferedImage rotatedSprite = this.rotateImage((BufferedImage) this.image, angle * 180);
            //this.image = rotatedSprite;
        }

        double incrementX = this.velX * (deltaTime - previousTime);
        double incrementY = this.velY * (deltaTime - previousTime);

        super.setTransform(this.x + incrementX, this.y + incrementY);
    }

    public void render(Graphics2D g2d) {
        int diameter = (int) (this.getDiameter() * this.imagePaddingFactor);
        int translationAdjustment = (diameter - this.getDiameter()) / 2;

        g2d.drawImage(this.image, (int) this.x - translationAdjustment, (int) this.y - translationAdjustment, diameter,
                diameter, null);
        Ellipse2D.Double obj = (Ellipse2D.Double) this.shape;
        obj.setFrame(this.x, this.y, this.getDiameter(), this.getDiameter());
    }
}

File: src/main/kotlin/com/ballgame/Fireball.kt
package com.ballgame

import java.awt.Graphics2D
import java.awt.Shape
import java.awt.geom.Ellipse2D
import java.awt.geom.Ellipse2D.*
import java.awt.image.BufferedImage
import kotlin.Double
import kotlin.math.atan

class Fireball(
	x: Int, y: Int, velX: Double, velY: Double, r: Int, sprites: List<BufferedImage>
) : Circle(x, y) {
	private val shape: Shape
	var velX: Double
	var velY: Double

	init {
		initSprites(sprites, 15, 1.6)
		this.velX = velX
		this.velY = velY
		shape = Double(this.x, this.y, diameter.toDouble(), diameter.toDouble())
	}

	fun getShape() = shape as Ellipse2D.Double

	fun rotateSprites() = sprites!!.indices.forEach {
		transformedSprites[it] = rotateImage(sprites!![it], atan(velX / -velY) * 180)
	}

	override fun update(deltaTime: Double, previousTime: Double) {
//		if (incrementImageFrame(deltaTime, previousTime)) image = rotateImage(image!!, atan(velX / -velY) * 180)
		super.setTransform(
			x + (velX * (deltaTime - previousTime)), y + (velY * (deltaTime - previousTime))
		)
	}

	override fun render(g2d: Graphics2D?) {
		val diameter = (diameter * imagePaddingFactor).toInt()
		val translationAdjustment = (diameter - this.diameter) / 2
		g2d!!.drawImage(
			image, x.toInt() - translationAdjustment, y.toInt() - translationAdjustment, diameter, diameter, null
		)
		(shape as Ellipse2D.Double).setFrame(
			x, y, this.diameter.toDouble(), this.diameter.toDouble()
		)
	}
}
